File: leetcode/cpp/graphs/lc_200_number_of_islands-sol-1.cpp
#include <vector>
#include <array>
#include <queue>

/**
 * https://leetcode.com/problems/number-of-islands/description/
 *
 * Time complexity: O(n * m) where n = grid row length, m = grid col length
 * Space complexity: O(n * m) where n = grid row length, m = grid col length
 */
class Solution {
public:
    // create a helper bfs function
    void bfs(
        const std::pair<int, int>& coords,
        const std::vector<std::vector<char>>& grid,
        std::vector<std::vector<bool>>& visited,
        const int& ROW_LEN,
        const int& COL_LEN
    ) {
        // setup our bfs queue
        std::queue<std::pair<int, int>> bfsQueue;
        // add the current coordinate to the bfs queue
        bfsQueue.push(coords);
        // mark the current coordinate as visited
        visited[coords.first][coords.second] = true;

        // setup our directions to search in
        const std::array<std::pair<int, int>, 4> directions{{
            {-1, 0},
            {1, 0},
            {0, 1},
            {0, -1}
        }};

        // standard bfs loop until empty
        while (!bfsQueue.empty()) {
            // get the first element from the bfs queue
            const auto [row, col] = bfsQueue.front();
            // remove the first element from the bfs queue
            bfsQueue.pop();

            // loop through all the directions
            for (const auto& [dx, dy]: directions) {
                // get the adjacent cell coordinates
                int adjRow = row+dx, adjCol = col+dy;

                // if:
                // 1) the adjacent cell is within grid bounds
                // 2) the adjacent cell has not been visited
                // 3) the adjacent cell contains a 1
                // then add it to the bfs queue and mark it as visited
                if (
                    (0 <= adjRow && adjRow < ROW_LEN) &&
                    (0 <= adjCol && adjCol < COL_LEN) &&
                    (!visited[adjRow][adjCol]) &&
                    (grid[adjRow][adjCol] == '1')
                ) {
                    bfsQueue.push(std::make_pair(adjRow, adjCol));
                    visited[adjRow][adjCol] = true;
                }
            }
        }
    }

    int numIslands(std::vector<std::vector<char>>& grid) {
        // get the row and column lengths of the grid
        const int ROW_LEN = grid.size();
        const int COL_LEN = grid[0].size();

        // init our visited 2d array as all false
        std::vector<std::vector<bool>> visited(ROW_LEN, std::vector<bool>(COL_LEN));

        // init our count
        int count = 0;

        // loop through all rows
        for (int row{0}; row < ROW_LEN; ++row) {
            // loop through all cols
            for (int col{0}; col < COL_LEN; ++col) {
                // if the cell is 1 and has not been visited before
                // run bfs on it and increment the count
                if (grid[row][col] == '1' && !visited[row][col]) {
                    bfs(std::make_pair(row, col), grid, visited, ROW_LEN, COL_LEN);
                    ++count;
                }
            }
        }

        // finally return the count
        return count;
    }
};

File: leetcode/graphs/lc_200_number_of_islands-sol-1.py
from typing import List


'''
https://leetcode.com/problems/number-of-islands/description/
'''
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        # get the dimensions of the grid        
        row_len = len(grid)
        col_len = len(grid[0])

        # setup the directions we can search in
        # for adjacent 1s to connect land forming an island
        directions = [(-1, 0), (1, 0), (0, 1), (0, -1)]

        # create a set to keep track of (row, col) pairs
        # we have already visited to form an island
        visited_ones = set()

        # our recursive depth first search function
        # used to explore all adjacent connected 1s
        # and mark them as visited, discovering an island
        def dfs(row, col):
            # first add the current (row, col) to the visited set
            visited_ones.add((row, col))
            
            # loop through all adjacent directions
            for dx, dy in directions:
                # get the adjacent cell
                adjacent_row = row+dx
                adjacent_col = col+dy

                # only run dfs on adjacent cell if
                # 1) adjacent cell is in grid bounds
                # 2) adjacent cell contains a '1'
                # 3) we have not visited this adjacent cell before
                if (0 <= adjacent_row < row_len and
                    0 <= adjacent_col < col_len and
                    grid[adjacent_row][adjacent_col] == '1' and
                    (adjacent_row, adjacent_col) not in visited_ones):
                    # run dfs on the adjacent row and col
                    dfs(adjacent_row, adjacent_col)

        # setup our variable to hold the count of the number of islands
        island_count = 0
        # loop through all rows in the grid
        for row in range(row_len):
            # loop through all cols in the grid
            for col in range(col_len):
                # only explore for an island if
                # 1) the cell contains a '1'
                # 2) we have not visited this cell before
                if (grid[row][col] == '1' and
                    (row, col) not in visited_ones):
                    # run our recursive dfs function to explore
                    # the island and mark it as visited
                    dfs(row, col)
                    # increment the island count
                    island_count += 1

        # finally return the total island count
        return island_count
